crates/ra_hir/src/from_source.rs

```diff
@@ -0,0 +1,217 @@
+use ra_db::{FileId, FilePosition};
+use ra_syntax::{
+    algo::find_node_at_offset,
+    ast::{self, AstNode, NameOwner},
+    SyntaxNode,
+};
+
+use crate::{
+    db::{AstDatabase, DefDatabase, HirDatabase},
+    ids::{AstItemDef, ConstId, EnumId, FunctionId, StaticId, StructId, TraitId, TypeAliasId},
+    name::AsName,
+    Const, Crate, Enum, EnumVariant, FieldSource, Function, HasSource, ImplBlock, Module,
+    ModuleSource, Source, Static, Struct, StructField, Trait, TypeAlias, Union, VariantDef,
+};
+
+pub trait FromSource {
+    type Ast;
+    fn from_source(db: &(impl DefDatabase + AstDatabase), src: Source<Self::Ast>) -> Option<Self>
+    where
+        Self: Sized;
+}
+
+impl FromSource for Struct {
+    type Ast = ast::StructDef;
+    fn from_source(db: &(impl DefDatabase + AstDatabase), src: Source<Self::Ast>) -> Option<Self> {
+        let id = StructId::from_source(db, src)?;
+        Some(Struct { id })
+    }
+}
+impl FromSource for Union {
+    type Ast = ast::StructDef;
+    fn from_source(db: &(impl DefDatabase + AstDatabase), src: Source<Self::Ast>) -> Option<Self> {
+        let id = StructId::from_source(db, src)?;
+        Some(Union { id })
+    }
+}
+impl FromSource for Enum {
+    type Ast = ast::EnumDef;
+    fn from_source(db: &(impl DefDatabase + AstDatabase), src: Source<Self::Ast>) -> Option<Self> {
+        let id = EnumId::from_source(db, src)?;
+        Some(Enum { id })
+    }
+}
+impl FromSource for Trait {
+    type Ast = ast::TraitDef;
+    fn from_source(db: &(impl DefDatabase + AstDatabase), src: Source<Self::Ast>) -> Option<Self> {
+        let id = TraitId::from_source(db, src)?;
+        Some(Trait { id })
+    }
+}
+impl FromSource for Function {
+    type Ast = ast::FnDef;
+    fn from_source(db: &(impl DefDatabase + AstDatabase), src: Source<Self::Ast>) -> Option<Self> {
+        let id = FunctionId::from_source(db, src)?;
+        Some(Function { id })
+    }
+}
+impl FromSource for Const {
+    type Ast = ast::ConstDef;
+    fn from_source(db: &(impl DefDatabase + AstDatabase), src: Source<Self::Ast>) -> Option<Self> {
+        let id = ConstId::from_source(db, src)?;
+        Some(Const { id })
+    }
+}
+impl FromSource for Static {
+    type Ast = ast::StaticDef;
+    fn from_source(db: &(impl DefDatabase + AstDatabase), src: Source<Self::Ast>) -> Option<Self> {
+        let id = StaticId::from_source(db, src)?;
+        Some(Static { id })
+    }
+}
+impl FromSource for TypeAlias {
+    type Ast = ast::TypeAliasDef;
+    fn from_source(db: &(impl DefDatabase + AstDatabase), src: Source<Self::Ast>) -> Option<Self> {
+        let id = TypeAliasId::from_source(db, src)?;
+        Some(TypeAlias { id })
+    }
+}
+impl FromSource for EnumVariant {
+    type Ast = ast::EnumVariant;
+    fn from_source(db: &(impl DefDatabase + AstDatabase), src: Source<Self::Ast>) -> Option<Self> {
+        EnumVariant::from_source_impl(db, src)
```

Comment:
This seems like a needless indirection, the `from_source_impl` is just below, let's inline it